pilot/pkg/networking/core/v1alpha3/listener_waypoint.go

```diff
@@ -301,127 +239,120 @@ func (lb *ListenerBuilder) buildWaypointInboundVIP(svcs map[host.Name]*model.Ser
-				l.FilterChains = append(l.FilterChains, tcpChain, httpChain)
-				l.FilterChainMatcher = match.NewAppProtocol(match.ProtocolMatch{
+				chains = append(chains, tcpChain, httpChain)
+				portMapper.Map[portString] = match.ToMatcher(match.NewAppProtocol(match.ProtocolMatch{
 					TCP:  match.ToChain(tcpName),
 					HTTP: match.ToChain(httpName),
-				})
+				}))
 			} else if port.Protocol.IsHTTP() {
 				// Otherwise, just insert HTTP/TCP
-				l.FilterChains = append(l.FilterChains, httpChain)
+				chains = append(chains, httpChain)
+				portMapper.Map[portString] = match.ToChain(httpChain.Name)
 			} else {
-				l.FilterChains = append(l.FilterChains, tcpChain)
+				chains = append(chains, tcpChain)
+				portMapper.Map[portString] = match.ToChain(tcpChain.Name)
 			}
-			listeners = append(listeners, l)
+		}
+		if len(portMapper.Map) > 0 {
+			cidr := util.ConvertAddressToCidr(svc.GetAddressForProxy(lb.node))
+			rangeMatcher := &matcher.IPMatcher_IPRangeMatcher{
+				Ranges: []*xds.CidrRange{{
+					AddressPrefix: cidr.AddressPrefix,
+					PrefixLen:     cidr.PrefixLen,
+				}},
+				OnMatch: match.ToMatcher(portMapper.Matcher),
+			}
+			ipMatcher.RangeMatchers = append(ipMatcher.RangeMatchers, rangeMatcher)
 		}
 	}
-	return listeners
-}
 
-// (many) Pod listener, `inbound||podip|port`. This is one per inbound pod. Will go through HCM if needed, in order to apply L7 policies (authz)
-// Note: we need both a pod listener and a VIP listener since we need to apply policies at different levels (routing vs authz).
-func (lb *ListenerBuilder) buildWaypointInboundPod(wls []WorkloadAndServices) []*listener.Listener {
-	listeners := []*listener.Listener{}
-	for _, wlx := range wls {
-		// Follow same logic as today, but no mTLS ever
-		wl := wlx.WorkloadInfo
-
-		// For each port, setup a match
-		// TODO: fake proxy is really bad. Should have these take in Workload or similar
-		instances := lb.Discovery.GetProxyServiceInstances(&model.Proxy{
-			Type:            model.SidecarProxy,
-			IPAddresses:     []string{wl.PodIP},
-			ConfigNamespace: wl.Namespace,
-			Metadata: &model.NodeMetadata{
-				Namespace: wl.Namespace,
-				Labels:    wl.Labels,
-			},
-		})
-		if len(instances) == 0 {
-			// TODO: Don't we need some passthrough mechanism? We will need ORIG_PORT but custom IP to implement that though
-			continue
-		}
-		wlBuilder := lb.WithWorkload(wl)
-		for _, port := range getPorts(instances) {
-			if port.Protocol == protocol.UDP {
-				continue
-			}
-			cc := inboundChainConfig{
-				clusterName: model.BuildSubsetKey(model.TrafficDirectionInboundPod, "", host.Name(wl.PodIP), port.Port),
-				port: ServiceInstancePort{
-					Name:       port.Name,
-					Port:       uint32(port.Port),
-					TargetPort: uint32(port.Port),
-					Protocol:   port.Protocol,
+	{
+		// Direct pod access chain.
+		directChain := &listener.FilterChain{
+			Name: "direct",
+			Filters: []*listener.Filter{{
+				Name: wellknown.TCPProxy,
```

Comment:
I am thinking eventually we are going to need to expand this to include policies, which also means sniffing and L7 potentially. WDYT?

(for followup)

Reply:
Yes, will need an HTTP direct chain. It's hard to reuse service building code without a service, but need a follow-up to share more code with service version.